### DISCUSSION TITLE
Calling fields by ID

### DISCUSSION THREAD

**Discussion by masgo** [johnkerl/miller#948] [Q&A] [answered]
Is there a way to call fields by id **or** by name within the same call? For example, I would like to do something like `then rename 1,newName`. This should rename column 1 in "newName". 

I saw [positional field names](https://miller.readthedocs.io/en/latest/reference-dsl-variables/#positional-field-names), but it seems they only work in the DSL, i.e. when called like `$[[[3]]]`

I am usually dealing with tsvlite or csv files. Sometimes they have long and complicated column names, sometimes even with non-english letters and other special characters etc. (example: `ID/Eintrag gelöscht (001)`) Miller works fine with that, but it can be quite difficult to address these fields, from within a script. I know that I can use curly braces or " (e.g., `"ID/Eintrag gelöscht (001)"` ). But now the charset of the bash script file and/or of the console becomes important. This is painful.

In these cases I use `--implicit-csv-header --headerless-csv-output` and adress the columns by id. Works fine, but longer scripts become hard to read. Using something like `then rename [[[1]]],newName` at the beginning would make life easier. But since this is outside of the DSL, miller searches for a field called [[[1]]]. Is there a way to make it search for the first column?

**Comment by aborruso** [comment, marked as answer]
I also think it can be useful https://github.com/johnkerl/miller/issues/310

**Comment by masgo** [comment]
Thanks, I have not seen the ticket. Your answer also answers my question. It's not possible right now.

**Comment by johnkerl** [comment]
@masgo @aborruso this is, at essence, the same issue as in https://github.com/johnkerl/miller/discussions/763.

Namely, the field names for verbs are just strings with no syntactic structure. In #763 if you have a field `item` which is a map with subfield `name` then in the DSL you can refer to `item.name` which will index `$*` with `"item"`, then index that with `"name"`, where the `"."` is parsed as a delimiter. But the `cut` verb (and all non-DSL verbs) looks for a single field named `"item.name"` (the `"."` is not parsed out as a delimiter).

Here, similarly, the DSL parses the `[[` and `]]` syntactically to give you the field indexed by `3`. But non-DSL verbs look for (and fail to find) a field whose name is literally `"[[3]]"`.

What we need is a grammar, and a parser, for field "names" in non-DSL verbs to give them the same kind of expressiveness they already enjoy in the DSL.
